### DISCUSSION TITLE
Respect .env.local file

### DISCUSSION THREAD

**Discussion by josias-r** [directus/directus#3932] [Feature Requests]
Usually with node packages if there's a `.env` file that is respected, the `.env.local` file is respected too (I'm thinking of [nextjs](https://github.com/vercel/next.js/blob/canary/packages/next-env/index.ts) and [cra](https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/config/env.js) as examples who have that logic in place).

I see no reason why directus should not read the `.local` env file as well 😁

I've had a couple of times where it would be a nice thing to have.
I'm happy to contribute, but it might be easier for someone of the active devs on this project, since it would only require like two lines of additional code 😛

> Note: I currently hack this by editing the compiled source files of directus if anyone else quickly wants to make use of this. Just edit `node_modules/directus/dist/env.js` and replace `dotenv_1.default.config();` with the following code:
```
const path_1 = require("path");
const fs_1 = require("fs");
const appDirectory = fs_1.realpathSync(process.cwd());
const resolveApp = relativePath => path_1.resolve(appDirectory, relativePath);

dotenv_1.default.config();
dotenv_1.default.config({
    path: resolveApp(".env.local"),
});

```

**Comment by josias-r** [comment]
@rijkvanzanten This seems to me like a quick and easy feature. I'd love to contribute myself if needed. From your contribution guidelines, I read that ideas need to be discussed first.
So a 👍🏼 or 👎🏼 would be nice for me to know if I can open a PR 😄 

Edit / Answer:
As discussed with @rijkvanzanten. It's a bit of an [80/20](https://docs.directus.io/contributing/introduction/#feature-requests) question, therefore currently not a priority. But for most cases, you can change the `CONFIG_PATH` variable for a local startup script in order to load different variables locally (i.e. `CONFIG_PATH=.env.local npx directus start`).

**Reply by rijkvanzanten** [reply]
You can already set the env file used by using the `DOTENV_CONFIG_PATH` environment variable, so that should cover the use case too right? Starting Directus with `DOTENV_CONFIG_PATH=.env.local npx directus start`

**Reply by josias-r** [reply]
@rijkvanzanten Oh wow, I did not know about that, maybe add this to the [docs](https://docs.directus.io/reference/environment-variables/)? 

There is the downside of being able to have both. So you can't add the .env.local to overwrite only the env variables that need to be changed.

I still see now downside of adding the two or three lines to load the local env though, but it does cover the ability to load different env variables for development, so thanks for your reply 😄

**Reply by rijkvanzanten** [reply]
> I still see now downside of adding the two or three lines to load the local env though

It's a bit of [an 80/20](https://docs.directus.io/contributing/introduction/#feature-requests) question. It's useful to have, but I'm wondering if it's useful enough to the majority of users. 

The ability to merge environment variable files locally raises a couple more questions: 

* Why `.env.local`, what if I want to call it `.env.dev` or otherwise? 
* What is the order in which it's merged? How do I specify which is the "base" and which is the add-on?
* Can you do more than one?

In this case, I'd say you can achieve what you're looking for by using [a `.js` config file](https://docs.directus.io/reference/config-files/#config-js), and handle the merging yourself. That way there's no confusion like that whatsoever 🤔

**Reply by josias-r** [reply]
I see. I guess as long as this discussion does not retrieve enough attention it's probably in the 20%.

I tried this though, and received the error `🚨 "DB_HOST" Environment Variable is missing.` even though it is specified inside .env.local:
```
"start": "cross-env DOTENV_CONFIG_PATH=.env.local directus start",
// or
"bootstrap": "cross-env DOTENV_CONFIG_PATH=.env.local directus bootstrap",
```
Are you sure the ENV var is supported?

Edit: Just figured it out myself by taking a look at [this file](https://github.com/directus/directus/blob/main/api/src/env.ts). Its CONFIG_PATH instead of DOTENV_CONFIG_PATH 😄

**Reply by rijkvanzanten** [reply]
ah yeah, my b

**Reply by jotatoledo** [reply]
oh man :smile:, considering the amount of features that directus offers out-of-the-box I assumed that an .env hierarchy would be among them. Would be nice IMO if the .env files could be resolved in the followin order, where they successively merge and overwrite the values of the previous:

1. .env
1. .env.local
1. .env.{env}
1. .env.{env}.local

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by romanlex** [comment]
Hello, maybe use dotenv-expand?

I have little another case where I can't use changed env at runtime
My custom start script to spawn a child process with `directus start`
Before spawn this process I use dotenv-expand

```js
const dotenvFiles = [
    Boolean(process.env.NODE_ENV) && path.join(pkgPath, `.env.${process.env.NODE_ENV}.local`),
    Boolean(process.env.NODE_ENV) && path.join(pkgPath, `.env.${process.env.NODE_ENV}`),
    Boolean(process.env.NODE_ENV) === false && path.join(pkgPath, `.env.development`),
    // Don't include `.env.local` for `test` environment
    // since normally you expect tests to produce the same
    // results for everyone
    process.env.NODE_ENV !== 'test' && path.join(pkgPath, `.env.local`),
    path.join(pkgPath, '.env'),
  ].filter(truthy)

  dotenvFiles.forEach((dotenvFile) => {
    if (fs.existsSync(dotenvFile)) {
      dotenvExpand.expand(
        dotenv.config({
          path: dotenvFile,
        })
      )
    }
  })
  const cp = spawn('bun', ['run', script], {
    cwd: pkg.path,
    env: { ...process.env, FORCE_COLOR: '1' },
  })
  ```
  
  It's not work because start script of directus load again .env file and ignore process.env

**Comment by oooth** [comment]
Using Docker Compose to define the following env files on your container:
```
# ...
  env_file:
    - .env
    - .env.foo
    - .env.foo.bar
```

Start command being `CONFIG_PATH="./get-env.js" npx directus start` and having `get-env.js` with the following (note that `env` param is a `process.env` object, which is ready as is thank to Docker Compose config):
```
export default function getEnv (env) {
  return Object.assign({}, env)
}
```
